Synchronization script for Gitlab

GG-Yanne opened the pull request:
Example script for synchronizing teams and members from a Gitlab instance to a GitGuardian workspace

rjouhann commented:
@GG-Yanne, can you rename the folder containing the script to `team-mapping-gitlab-gitguardian` to match @alustrat-gg https://github.com/GitGuardian/ggtools/pull/24?

Also, could you add a readme and sync with @alustrat-gg so you both use the same template? The readme would contain installation steps including requirements, few examples of use, etc...

GG-Yanne commented:
> @GG-Yanne, can you rename the folder containing the script to `team-mapping-gitlab-gitguardian` to match @Alustrat-gg #24?

I moved the script files to the folder `team-mapping-gitlab-gitguardian` 

> Also, could you add a readme and sync with @Alustrat-gg so you both use the same template? The readme would contain installation steps including requirements, few examples of use, etc...

I adapted his readme here 👍

rjouhann commented:
Hello @GG-Yanne, 

Same comment I made to @Alustrat-gg:

1. Could we add an option to notify or not users of changes, invitations, or removal from teams?
2. Are you excluding users with elevated permissions (manager, owner) to avoid redundant access grants?
3. Could you explain in the readme how we map the GitHub team and GitGuardian team and how we handle nested teams?
4. Should we add a "mini" changelog in the readme in case we modify the script in the future as well as a version?

Thanks!

GG-Yanne commented:
> Hello @GG-Yanne,
> 
> Same comment I made to @Alustrat-gg:
> 
> 1. Could we add an option to notify or not users of changes, invitations, or removal from teams?

I added the option in the config file and mentioned it in the README

> 2. Are you excluding users with elevated permissions (manager, owner) to avoid redundant access grants?

There is an option that users can opt-in to exclude admin users

> 3. Could you explain in the readme how we map the GitHub team and GitGuardian team and how we handle nested teams?

I added explanations around how we handle nested teams, tl;dr we fetch users as opposed to groups so we get users' groups and we only use the name of the group not its full path

> 4. Should we add a "mini" changelog in the readme in case we modify the script in the future as well as a version?

I am syncing with Arnaud to see what should we do

wasshana-gg commented on team-mapping-gitlab-gitguardian/README.md:
```diff
@@ -0,0 +1,60 @@
+
+Teams in GitGuardian will be created based on the name (not the **full path**) of every user's group.
+
+This means that if a user is in `parent-group / child-group`, he will be added to the team `child-group` in GitGuardian.
```
Does this mean that we build the team based on the subgroup in this example? Is it based on the last group in the path? 
Can we clarify this point so that it's clear on what the expected behaviour is? 
What's the depth ? Group / subgroup / sub sub group / sub sub sub group / ?

GG-Yanne replied:
I clarified this point, in your case users would be added to the team `sub sub sub group`, it will always be the deepest group name.

Let me know if it's clearer 👍

wasshana-gg commented on team-mapping-gitlab-gitguardian/README.md:
```diff
@@ -0,0 +1,60 @@
+
+### Invoking
+
+Upon invocation, the script will sync teams and their perimeters from GitHub to GitGuardian. It can be invoked like this:
```
Upon invocation, the script will sync teams and their perimeters from Gitlab ~GitHub~ to GitGuardian. It can be invoked like this:

GG-Yanne replied:
Fixed this 😅

QAGitGuardianAccount commented:
@GG-Yanne can you also add a reference to the GitLab script on the README at the root of the repo? https://github.com/GitGuardian/ggtools/blob/main/README.md (like we already have for GitHub)

rjouhann commented:
@GG-Yanne can you also add a reference to the GitLab script on the README at the root of the repo? https://github.com/GitGuardian/ggtools/blob/main/README.md (like we already have for GitHub)

rjouhann commented on team-mapping-gitlab-gitguardian/README.md:
```diff
@@ -0,0 +1,60 @@
+# Team mapping: Gitlab to GitGuardian
```
@GG-Yanne can you also add a reference to the GitLab script on the README at the root of the repo? https://github.com/GitGuardian/ggtools/blob/main/README.md (like we already have for GitHub)

GG-Yanne replied:
Sure, It's done 👍

wasshana-gg commented on team-mapping-gitlab-gitguardian/gitguardian_client.py:
```diff
@@ -0,0 +1,301 @@
+    member: Member,
+    team: Team,
+    is_team_leader: bool = False,
+    incident_permission: IncidentPermission = IncidentPermission.VIEW,
```
Do we know why .VIEW?
Is this a configurable value on execution ? 
It doesn't seem to be an argument that can be passed. 
Shouldn't the users specify whether all the members should have .VIEW .EDIT .FULL_ACCESS 
From a least privilege .VIEW makes sense, but most of the team members that are created have permissions to EDIT incidents.

GG-Yanne replied:
I added the ability to configure this value, it is documented in the README.

For `.VIEW` I just went with the least privileged permission, we can go with `.EDIT` as default if it is the most common 👍

GG-Yanne replied:
Indeed, I changed it in the last push force 👍

wasshana-gg reviewed (commented):
Some questions on my end
